### PR TITLE
added tekelek oiltank

### PR DESCRIPTION
**Pull request by J2thatsme**

### CODE REVIEW

File: types/tekelek/TEK766/uplink.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    var min1 = Bits.bitsToUnsigned(bits.substr(88, 8));
+    var min2 = Bits.bitsToUnsigned(bits.substr(96, 8));
+
+    data.meassurements = (min1 * 256) + min2;
+    data.transmitPeriods = Bits.bitsToUnsigned(bits.substr(104, 8));
+    data = getValues(bits, data, 112, 2, 'none');
+    topic = "lifecycle";
+  } else if (type == 67) {
+    //16 Reserved
+    topic = "paramRead";
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:115]: why so many different topics with the same structure?

**Review by mariusgiger — CHANGES_REQUESTED** [review]
the uplink decoder is still a bit confusing, please refactor

File: types/tekelek/TEK766/current.schema.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      "title": "Limit 2",
+      "type": "boolean",
+      "description": "True if configured limit 2 is exceeded"
+    },
+    "lim3": {
+      "title": "Limit 3",
+      "type": "boolean",
+      "description": "True if configured limit 3 is exceeded"
+    },
+    "SRSSI1": {
```
[the remark this answers is not in the document]
Comment (J2thatsme) [reply types/tekelek/TEK766/current.schema.json:36]: There can be multiple datapoints with this value/key in one payload.

File: types/tekelek/TEK766/current.schema.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      "title": "Sonic Received Signal Strength Indicator",
+      "type": "integer",
+      "description": "The SRSSI is simply an integer number between 1 (low level) and 10 (expected level) given to the expected strength of the echo reflected to the transducer.",
+      "max": 10,
+      "min": 1
+    },
+    "SRC1": {
```
[the remark this answers is not in the document]
Comment (J2thatsme) [reply types/tekelek/TEK766/current.schema.json:43]: There can be multiple datapoints with this value/key in one payload.

File: types/tekelek/TEK766/lifecycle.schema.json [hunk 4]
```diff
@@ -0,0 +1,97 @@
+{
+  "$id": "https://akenza.io/tekelek/tek766/lifecycle.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Lifecycle",
+  "type": "object",
+  "properties": {
+    "batteryLevel": {
```
[the remark this answers is not in the document]
Comment (J2thatsme) [reply types/tekelek/TEK766/lifecycle.schema.json:7]: yes this one should be the standart from now

File: types/tekelek/TEK766/current.schema.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mariusgiger) [reply types/tekelek/TEK766/current.schema.json:43]: ok but those are not in the schema?

File: types/tekelek/TEK766/lifecycle.schema.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (mariusgiger) [reply types/tekelek/TEK766/lifecycle.schema.json:7]: ok, please correct it in the other places where still batteryPercent was used

File: types/tekelek/TEK766/alarm_history.schema.json [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{
+  "$id": "https://akenza.io/tekelek/tek766/alarm_history.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Alarm history",
+  "type": "object",
+  "properties": {
+    "limit1": {
```
Comment (mariusgiger) [inline types/tekelek/TEK766/alarm_history.schema.json:7]: rename to limit1Exceeded?

File: types/tekelek/TEK766/alarm_history.schema.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      "unit": "°C",
+      "type": "number",
+      "description": "The temperature in °C.",
+      "minimum": -3276.5,
+      "maximum": 3276.5
+    },
+    "srssi": {
+      "title": "Sonic Received Signal Strength Indicator",
+      "type": "integer",
+      "description": "The SRSSI is simply an integer number between 1 (low level) and 10 (expected level) given to the expected strength of the echo reflected to the transducer.",
```
Comment (mariusgiger) [inline types/tekelek/TEK766/alarm_history.schema.json:39]: remove simply

File: types/tekelek/TEK766/lifecycle.schema.json [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      "description": "Contact reason",
+      "type": "string",
+      "enum": ["Reset", "Scheduled", "Manual", "Activation"]
+    },
+    "systemRequestReset": {
+      "title": "System request reset",
+      "description": "Defines the reason for the last reset",
+      "type": "string",
+      "enum": [
+        "Power on reset",
```
Comment (mariusgiger) [inline types/tekelek/TEK766/lifecycle.schema.json:36]: enum values should conform to Java enums POWER_ON_RESET

File: types/tekelek/TEK766/alarm_history.schema.json (hunk 6 above)
Comment (mariusgiger) [reply types/tekelek/TEK766/alarm_history.schema.json:39]: do this everywhere

File: types/tekelek/TEK766/meta.json [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  "author": "Akenza AG",
+  "firmwareVersion": "V1.0.0",
+  "loraDeviceClass": "A",
+  "availableSensors": [
+    "Distance",
+    "Temperature"
+  ],
+  "outputTopics": [
+    "current",
+    "measurement",
```
Comment (mariusgiger) [inline types/tekelek/TEK766/meta.json:16]: missing schemas

File: types/tekelek/TEK766/param_read.schema.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+  "$id": "https://akenza.io/tekelek/tek766/param_read.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Param read",
+  "type": "object",
+  "properties": {}
```
Comment (mariusgiger) [inline types/tekelek/TEK766/param_read.schema.json:6]: missing properties

File: types/tekelek/TEK766/uplink.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
+function getUllage(ull1, ull2) {
+  return ull1 * 256 + ull2;
+}
+function getTemperature(temp) {
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:4]: add a comment to explain this calculation

File: types/tekelek/TEK766/uplink.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
+function getUllage(ull1, ull2) {
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:1]: add a comment to explain this calculation

File: types/tekelek/TEK766/uplink.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    let temp = Bits.bitsToUnsigned(bits.substr(48, 8));
+    data.temperature = getTemperature(temp);
+    data.srssi = Bits.bitsToUnsigned(bits.substr(56, 4));
+    data.src = Bits.bitsToUnsigned(bits.substr(60, 4));
+
+    if (type === 16) {
+      emit("sample", { data, topic: "current" });
+
+      topic = "measurement_history";
+      data = {};
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:37]: don't reuse variables

File: types/tekelek/TEK766/uplink.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+function consume(event) {
+  const payload = event.data.payload_hex;
+  const bits = Bits.hexToBits(payload);
+  let data = {};
+  let topic = "default";
+  const type = Bits.bitsToUnsigned(bits.substr(0, 8));
+  // 8-16 prodID
+  if (type === 16 || type === 69) {
+    data.limit1 = !!Number(bits.substr(23, 1));
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:20]: why is limit 1 at 23?

File: types/tekelek/TEK766/uplink.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        data[`ullage${i}`] = getUllage(ull1, ull2);
+
+        temp = Bits.bitsToUnsigned(bits.substr((pointer += 8), 8));
+        data[`temperature${i}`] = getTemperature(temp);
+        data[`srssi${i}`] = Bits.bitsToUnsigned(bits.substr((pointer += 8), 4));
+        data[`src${i}`] = Bits.bitsToUnsigned(bits.substr((pointer += 4), 4));
+        pointer += 4;
+      }
+    } else {
+      emit("sample", { data, topic: "current_alarm" });
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:51]: why emit here?

File: types/tekelek/TEK766/uplink.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    let ull2 = Bits.bitsToUnsigned(bits.substr(40, 8));
+    data.ullage = getUllage(ull1, ull2);
+
+    let temp = Bits.bitsToUnsigned(bits.substr(48, 8));
+    data.temperature = getTemperature(temp);
+    data.srssi = Bits.bitsToUnsigned(bits.substr(56, 4));
+    data.src = Bits.bitsToUnsigned(bits.substr(60, 4));
+
+    if (type === 16) {
+      emit("sample", { data, topic: "current" });
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:34]: why emit here?

File: types/tekelek/TEK766/uplink.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+      temp = Bits.bitsToUnsigned(bits.substr(80, 8));
+      data.temperature = getTemperature(temp);
+      data.srssi = Bits.bitsToUnsigned(bits.substr(88, 4));
+      data.src = Bits.bitsToUnsigned(bits.substr(96, 4));
+    }
+  } else if (type === 48) {
+    topic = "lifecycle";
+    // 16 reserved
+    data.hardwareID = Bits.bitsToUnsigned(bits.substr(24, 8));
+    data.fmVersion = `${Bits.bitsToUnsigned(
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:68]: extract to variables before concatenating the string

File: types/tekelek/TEK766/uplink.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+      bits.substr(32, 8),
+    )}.${Bits.bitsToUnsigned(bits.substr(40, 8))}`;
+    data.manualContact = Bits.bitsToUnsigned(bits.substr(54, 2));
+    data.systemRequestReset = Bits.bitsToUnsigned(bits.substr(51, 3));
+    data.aktive = Bits.bitsToUnsigned(bits.substr(50, 1));
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:73]: aktive? active

File: types/tekelek/TEK766/uplink.js [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+    // Reserved 56
+    data.RSSI = -Bits.bitsToUnsigned(bits.substr(64, 8));
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:75]: either always use small letters srssi or always capital letters RSSI

File: types/tekelek/TEK766/uplink.js [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    data.srssi = Bits.bitsToUnsigned(bits.substr(136, 4));
+    data.src = Bits.bitsToUnsigned(bits.substr(140, 4));
+  } else if (type === 67) {
+    // 16 Reserved
+    topic = "param_read";
+  } else if (type === 71) {
+    topic = "diagnostic_read";
+  }
+
+  emit("sample", { data, topic });
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:100]: better move this up to the respective cases

File: types/tekelek/TEK766/uplink.js [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    data.systemRequestReset = Bits.bitsToUnsigned(bits.substr(51, 3));
+    data.aktive = Bits.bitsToUnsigned(bits.substr(50, 1));
+    // Reserved 56
+    data.RSSI = -Bits.bitsToUnsigned(bits.substr(64, 8));
+    // Reserved 72
+    data.battery = Bits.bitsToUnsigned(bits.substr(80, 8));
+
+    const min1 = Bits.bitsToUnsigned(bits.substr(88, 8));
+    const min2 = Bits.bitsToUnsigned(bits.substr(96, 8));
+    data.meassurements = min1 * 256 + min2;
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:81]: typo meassurements -> measurements

File: types/tekelek/TEK766/uplink.js [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+    data.fmVersion = `${Bits.bitsToUnsigned(
+      bits.substr(32, 8),
+    )}.${Bits.bitsToUnsigned(bits.substr(40, 8))}`;
+    data.manualContact = Bits.bitsToUnsigned(bits.substr(54, 2));
+    data.systemRequestReset = Bits.bitsToUnsigned(bits.substr(51, 3));
+    data.aktive = Bits.bitsToUnsigned(bits.substr(50, 1));
+    // Reserved 56
+    data.RSSI = -Bits.bitsToUnsigned(bits.substr(64, 8));
+    // Reserved 72
+    data.battery = Bits.bitsToUnsigned(bits.substr(80, 8));
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.js:77]: batteryLevel?

File: types/tekelek/TEK766/uplink.spec.js [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+          assert.equal(value.data.ullage3, 276);
+          assert.equal(value.data.temperature3, -14);
+          assert.equal(value.data.src3, 4);
+          assert.equal(value.data.srssi3, 7);
+          validate(value.data, measurementHistorySchema, { throwError: true });
+        }
+      });
+
+      consume(data);
+    });
```
Comment (mariusgiger) [inline types/tekelek/TEK766/uplink.spec.js:90]: add tests for the other payload types

**Review by mariusgiger — CHANGES_REQUESTED** [review]
thanks for the PR - looks already much better, some comments remain

File: types/tekelek/TEK766/uplink.js (hunk 13 above)
Comment (J2thatsme) [reply types/tekelek/TEK766/uplink.js:20]: skipping unused bits

**Review by mariusgiger — APPROVED** [review]

**Pull request merged**